### ISSUE TITLE
[snowflake] Commit with wrong session when merge 

### ISSUE THREAD

**Issue by troilus-canva** [fishtown-analytics/dbt#2820]
### Describe the bug
When running incremental update with DBT, the transaction started for merge is not committed after. Though there is commit command running after, but with different session_id.

### Steps To Reproduce
Not 100% how to reproduce. But with long enough time to run the merge query, and couple of post hooks, it ends up with committing with wrong session_id.

### Expected behavior
Commit with the same session_id and release the lock.

### Screenshots and log output
<img width="1099" alt="Screen Shot 2020-10-08 at 1 55 51 pm" src="https://user-images.githubusercontent.com/55678940/95424900-020b0680-0976-11eb-83da-9a86cf81f3c5.png">
<img width="1597" alt="Screen Shot 2020-10-08 at 3 00 33 pm" src="https://user-images.githubusercontent.com/55678940/95425564-0edc2a00-0977-11eb-9a53-b6b2b9559006.png">

### System information
**Which database are you using dbt with?**
- [ ] postgres
- [ ] redshift
- [ ] bigquery
- [x ] snowflake
- [ ] other (specify: ____________)

**The output of `dbt --version`:**
```
installed version: 0.17.0
   latest version: 0.18.0

Your version of dbt is out of date! You can find instructions for upgrading here:
https://docs.getdbt.com/docs/installation

Plugins:
  - bigquery: 0.17.0
  - snowflake: 0.17.0
  - redshift: 0.17.0
  - postgres: 0.17.0
```

**The operating system you're using:**
macos 10.15.6

**The output of `python --version`:**
Python 3.6.8 :: Anaconda, Inc.

**Comment by jtcohen6** [comment]
Hey @troilus-canva, first off, very cool github handle.

I think this will also be an issue with dbt versions 0.17.2+, since we now close Snowflake connections after use (#2650), such that post hooks (and likely `commit`, too) run inside of a different session.

What are the implications here? Does the `merge` fail to delete/update/insert records? Is there a lock on the table that prevents subsequent queries from completing?

**Comment by troilus-canva** [comment]
The `merge` is finished successfully, as you can see from the status as well. But the record is not inserted or updated. The lock will also prevents subsequent DML operations, unless it's expired or released manually.

**Comment by jtcohen6** [comment]
Got it, okay, definitely concerning.

I tried to put together a test case for this, but I haven't been able to reproduce what you're seeing. Here's what I've got right now (running with dbt v0.18.1-rc1):
```sql
{{ config(
    materialized = 'incremental', 
    unique_key = 'id',
    post_hook = "select 'this is my post hook'"
) }}

select

    1 as id,
    not {{ flags.FULL_REFRESH }} as did_my_merge_succeed,
    current_timestamp as current_ts
```
```
dbt run --full-refresh && dbt run
```

```
select * from dbt_jcohen.my_model
```
ID | DID_MY_MERGE_SUCCEED | CURRENT_TS
-- | -- | --
1 | true | 2020-10-08 12:13:47

<img width="1402" alt="Screen Shot 2020-10-08 at 12 15 06 PM" src="https://user-images.githubusercontent.com/13897643/95485600-fb719480-095f-11eb-820d-95bd68dfe63a.png">

**Comment by troilus-canva** [comment]
> I tried to put together a test case for this, but I haven't been able to reproduce what you're seeing. Here's what I've got right now (running with dbt v0.18.1-rc1):

yeah, I can't reproduce it all the time either. But it seems to related to the time it took to merge. The ones failed for me usually tooks more than 40 mins to merge.

**Comment by jtcohen6** [comment]
Ah that's helpful context. Do you by any chance have the [`client_session_keep_alive`](https://docs.getdbt.com/reference/warehouse-profiles/snowflake-profile/#client_session_keep_alive) config enabled in your profile?

**Comment by troilus-canva** [comment]
> Ah that's helpful context. Do you by any chance have the [`client_session_keep_alive`](https://docs.getdbt.com/reference/warehouse-profiles/snowflake-profile/#client_session_keep_alive) config enabled in your profile?

Thanks. I'll have this enabled and see if it can fix the issue.
